File: Sequence/cds2prot.pl
#!/usr/bin/perl -w
use strict;

unless ( $#ARGV >= 0 ){
	print "Usage: perl cds2prot.pl cds.fa > prot.fa\n";
	exit;
}	
my ($id,$protein) = ('','');
my ($i,$j)=(0,0);
my @id_a = ();
my %seq = ();

#reading Seq
foreach my $file (@ARGV){
	open SEQ, "<","$file" or die "Can't open file \"$file\" : $!\n";
	while (<SEQ>){
		chomp;
		if (/^>/){
			$id = $_;
			push @id_a,$_;
		} else {
			$_ = uc($_);
			$_ =~ s/U/T/g;
			$seq{$id} .= $_;
		}
	}
	close SEQ;
}

#translate
foreach my $key (@id_a){
	$j++;

#The 3 line below use to find the sequence id with bad codon.
#$id = (split /\s+/,$key)[0];
#$id =~ s/^>//;
#print STDERR "$j. $id\n";
	for(my $i=0;$i<(length($seq{$key})-2);$i+=3){
		my $codon=substr($seq{$key},$i,3);
		$protein.=&codon2aa($codon);
  	}
	$protein =~ s/(.{80})/$1\n/g;
  	print "$key\n$protein\n";
    	$protein = "";
}
print STDERR "Completed.\n\n";

#codon to amino acid
sub codon2aa{
	my ($codon)=@_;
my(%g)=('TCA'=>'S','TCC'=>'S','TCG'=>'S','TCT'=>'S','TTC'=>'F','TTT'=>'F','TTA'=>'L','TTG'=>'L','TAC'=>'Y','TAT'=>'Y','TAA'=>'*','TAG'=>'*','TGC'=>'C','TGT'=>'C','TGA'=>'*','TGG'=>'W','CTA'=>'L','CTC'=>'L','CTG'=>'L','CTT'=>'L','CCA'=>'P','CCC'=>'P','CCG'=>'P','CCT'=>'P','CAC'=>'H','CAT'=>'H','CAA'=>'Q','CAG'=>'Q','CGA'=>'R','CGC'=>'R','CGG'=>'R','CGT'=>'R','ATA'=>'I','ATC'=>'I','ATT'=>'I','ATG'=>'M','ACA'=>'T','ACC'=>'T','ACG'=>'T','ACT'=>'T','AAC'=>'N','AAT'=>'N','AAA'=>'K','AAG'=>'K','AGC'=>'S','AGT'=>'S','AGA'=>'R','AGG'=>'R','GTA'=>'V','GTC'=>'V','GTG'=>'V','GTT'=>'V','GCA'=>'A','GCC'=>'A','GCG'=>'A','GCT'=>'A','GAC'=>'D','GAT'=>'D','GAA'=>'E','GAG'=>'E','GGA'=>'G','GGC'=>'G','GGG'=>'G','GGT'=>'G','AAN'=>'X','GAN'=>'X','CAN'=>'X','TAN'=>'X','NAA'=>'X','NAG'=>'X','NAC'=>'X','NAT'=>'X','NAN'=>'X','AGN'=>'X','GGN'=>'X','CGN'=>'X','TGN'=>'X','NGA'=>'X','NGG'=>'X','NGC'=>'X','NGT'=>'X','NGN'=>'X','ACN'=>'X','GCN'=>'X','CCN'=>'X','TCN'=>'X','NCA'=>'X','NCG'=>'X','NCC'=>'X','NCT'=>'X','NCN'=>'X','ATN'=>'X','GTN'=>'X','CTN'=>'X','TTN'=>'X','NTA'=>'X','NTG'=>'X','NTC'=>'X','NTT'=>'X','NTN'=>'X','ANA'=>'X','ANG'=>'X','ANC'=>'X','ANT'=>'X','ANN'=>'X','GNA'=>'X','GNG'=>'X','GNC'=>'X','GNT'=>'X','GNN'=>'X','CNA'=>'X','CNG'=>'X','CNC'=>'X','CNT'=>'X','CNN'=>'X','TNA'=>'X','TNG'=>'X','TNC'=>'X','TNT'=>'X','TNN'=>'X','NNA'=>'X','NNG'=>'X','NNC'=>'X','NNT'=>'X','NNN'=>'X');
	if(exists $g{$codon}){
		return $g{$codon};
	}else{
		print STDERR "\nBad codon \"$codon\"!!\n";
		print STDERR "Not Completed!\n";
		exit;
	}
}

File: Sequence/cds2prot.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-

from __future__ import print_function
import sys
import argparse

TABLES = {
    'standard': {
        'TTT':'F','TTC':'F','TTA':'L','TTG':'L',
        'TCT':'S','TCC':'S','TCA':'S','TCG':'S',
        'TAT':'Y','TAC':'Y','TAA':'*','TAG':'*',
        'TGT':'C','TGC':'C','TGA':'*','TGG':'W',
        'CTT':'L','CTC':'L','CTA':'L','CTG':'L',
        'CCT':'P','CCC':'P','CCA':'P','CCG':'P',
        'CAT':'H','CAC':'H','CAA':'Q','CAG':'Q',
        'CGT':'R','CGC':'R','CGA':'R','CGG':'R',
        'ATT':'I','ATC':'I','ATA':'I','ATG':'M',
        'ACT':'T','ACC':'T','ACA':'T','ACG':'T',
        'AAT':'N','AAC':'N','AAA':'K','AAG':'K',
        'AGT':'S','AGC':'S','AGA':'R','AGG':'R',
        'GTT':'V','GTC':'V','GTA':'V','GTG':'V',
        'GCT':'A','GCC':'A','GCA':'A','GCG':'A',
        'GAT':'D','GAC':'D','GAA':'E','GAG':'E',
        'GGT':'G','GGC':'G','GGA':'G','GGG':'G',
    },
    # Vertebrate mitochondrial: TGA=W, AGA/AGG=*, ATA=M
    'vertmt': {
        'TTT':'F','TTC':'F','TTA':'L','TTG':'L',
        'TCT':'S','TCC':'S','TCA':'S','TCG':'S',
        'TAT':'Y','TAC':'Y','TAA':'*','TAG':'*',
        'TGT':'C','TGC':'C','TGA':'W','TGG':'W',
        'CTT':'L','CTC':'L','CTA':'L','CTG':'L',
        'CCT':'P','CCC':'P','CCA':'P','CCG':'P',
        'CAT':'H','CAC':'H','CAA':'Q','CAG':'Q',
        'CGT':'R','CGC':'R','CGA':'R','CGG':'R',
        'ATT':'I','ATC':'I','ATA':'M','ATG':'M',
        'ACT':'T','ACC':'T','ACA':'T','ACG':'T',
        'AAT':'N','AAC':'N','AAA':'K','AAG':'K',
        'AGT':'S','AGC':'S','AGA':'*','AGG':'*',
        'GTT':'V','GTC':'V','GTA':'V','GTG':'V',
        'GCT':'A','GCC':'A','GCA':'A','GCG':'A',
        'GAT':'D','GAC':'D','GAA':'E','GAG':'E',
        'GGT':'G','GGC':'G','GGA':'G','GGG':'G',
    },
    # Invertebrate mitochondrial: TGA=W, AGA/AGG=S, ATA=M
    'invertmt': {
        'TTT':'F','TTC':'F','TTA':'L','TTG':'L',
        'TCT':'S','TCC':'S','TCA':'S','TCG':'S',
        'TAT':'Y','TAC':'Y','TAA':'*','TAG':'*',
        'TGT':'C','TGC':'C','TGA':'W','TGG':'W',
        'CTT':'L','CTC':'L','CTA':'L','CTG':'L',
        'CCT':'P','CCC':'P','CCA':'P','CCG':'P',
        'CAT':'H','CAC':'H','CAA':'Q','CAG':'Q',
        'CGT':'R','CGC':'R','CGA':'R','CGG':'R',
        'ATT':'I','ATC':'I','ATA':'M','ATG':'M',
        'ACT':'T','ACC':'T','ACA':'T','ACG':'T',
        'AAT':'N','AAC':'N','AAA':'K','AAG':'K',
        'AGT':'S','AGC':'S','AGA':'S','AGG':'S',
        'GTT':'V','GTC':'V','GTA':'V','GTG':'V',
        'GCT':'A','GCC':'A','GCA':'A','GCG':'A',
        'GAT':'D','GAC':'D','GAA':'E','GAG':'E',
        'GGT':'G','GGC':'G','GGA':'G','GGG':'G',
    }
}

def parse_args():
    p = argparse.ArgumentParser(
        description="Translate CDS FASTA to protein (Python 2.7 compatible)."
    )
    p.add_argument('fastas', nargs='+', help='Input CDS FASTA file(s)')
    p.add_argument('--table', '-g', default='standard',
                   choices=['standard', 'vertmt', 'invertmt'],
                   help='Genetic code (default: standard)')
    p.add_argument('--trim_at_stop', action='store_true',
                   help='Truncate at first stop (*)')
    p.add_argument('--drop_terminal_stop', action='store_true',
                   help='Drop a single terminal * if present')
    p.add_argument('--frame', type=int, default=1, choices=[1,2,3],
                   help='Reading frame (default: 1)')
    p.add_argument('--width', type=int, default=80,
                   help='Wrap width (default: 80)')
    p.add_argument('--clean_id', action='store_true',
                   help='Keep only first token in header')
    return p.parse_args()

def read_fastas(paths):
    order = []
    seqs = {}
    for path in paths:
        with open(path, 'r') as fh:
            hdr = None
            for raw in fh:
                line = raw.rstrip('\n\r')
                if not line:
                    continue
                if line.startswith('>'):
                    hdr = line
                    order.append(hdr)
                    if hdr not in seqs:
                        seqs[hdr] = []
                else:
                    line = line.upper().replace('U', 'T')
                    line = ''.join(line.split())
                    if hdr is not None:
                        seqs[hdr].append(line)
    # join
    for k in seqs:
        seqs[k] = ''.join(seqs[k])
    return order, seqs

def translate(seq, code, trim_at_stop):
    """Translate DNA (ACGT + IUPAC) -> protein.
       Any codon containing chars outside A/C/G/T => 'X'."""
    L = len(seq)
    prot = []
    i = 0
    while i + 2 < L:
        codon = seq[i:i+3]
        i += 3
        if not codon or len(codon) < 3:
            break
        if _has_non_acgt(codon):
            aa = 'X'
        else:
            aa = code.get(codon, 'X')  # ACGT-only but unexpected => X
        if trim_at_stop and aa == '*':
            break
        prot.append(aa)
    return ''.join(prot)

def _has_non_acgt(codon):
    for ch in codon:
        if ch not in ('A','C','G','T'):
            return True
    return False

def wrap_seq(s, w):
    if w <= 0:
        return s
    out = []
    for i in range(0, len(s), w):
        out.append(s[i:i+w])
    return '\n'.join(out)

def main():
    args = parse_args()
    code = TABLES[args.table]
    if args.width < 1:
        sys.stderr.write('--width must be >= 1\n')
        sys.exit(2)

    order, seqs = read_fastas(args.fastas)

    for hdr in order:
        dna = seqs.get(hdr, '')
        if not dna:
            continue

        # frame offset
        if args.frame > 1:
            if len(dna) < (args.frame - 1):
                sys.stderr.write('%s: sequence shorter than frame offset; skipping\n' % hdr)
                continue
            dna = dna[args.frame - 1:]

        L = len(dna)
        rem = L % 3
        if rem != 0:
            sys.stderr.write('%s length %d not multiple of 3; trailing %d nt ignored\n' % (hdr, L, rem))

        prot = translate(dna, code, args.trim_at_stop)

        if (not args.trim_at_stop) and args.drop_terminal_stop and prot.endswith('*'):
            prot = prot[:-1]

        out_hdr = hdr
        if args.clean_id:
            h = hdr[1:] if hdr.startswith('>') else hdr
            h = h.split()[0]
            out_hdr = '>' + h

        print(out_hdr)
        print(wrap_seq(prot, args.width))

if __name__ == '__main__':
    main()
